### PR TITLE
[SPARK-47793][TEST][FOLLOWUP] Fix flaky test for Python data source exactly once.

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Fix the flakiness in python streaming source exactly once test. The last executed batch may not be recorded in query progress, which cause the expected rows doesn't match. This fix takes the uncompleted batch into account and relax the condition

### Why are the changes needed?
Fix flaky test.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Test change.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a follow-up, @chaoqin-li1123 .

I revised the PR title accordingly to Apache Spark convention.

**Comment by dongjoon-hyun** [comment]
cc @HeartSaVioR and @allisonwang-db from #45977

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala [hunk 1]
```diff
@@ -326,8 +326,11 @@ class PythonStreamingDataSourceSuite extends PythonDataSourceSuiteBase {
         lastBatch = q.lastProgress.batchId.toInt
       }
       assert(lastBatch > 20)
+      val rowCount = spark.read.format("json").load(outputDir.getAbsolutePath).count()
+      // There may be one uncommitted batch that is not recorded in query progress.
+      assert(rowCount == 2 * lastBatch + 2 || rowCount == 2 * lastBatch + 4)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:331]: Is this a correct way to fix the root cause of flakiness?

Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:331]: Unfortunately there is no graceful way to shutdown a streaming query. I can't think of any alternative.

Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:331]: Can we explain the `2 * lastBatch + 2` and `2 * lastBatch + 4` in the comment here? Just curious why can't we provide an upper bound for the row count here?

Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala]: More comment added. We can check an upper bound here, but it is not really meaningful.

**Review by allisonwang-db — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
It looks fine if this is the best.

**Comment by dongjoon-hyun** [comment]
Could you do the final review and sign-off, please, @HyukjinKwon ?

**Comment by dongjoon-hyun** [comment]
Let me bring this first for further monitoring. Thank you, @chaoqin-li1123 and @allisonwang-db .
Merged to master.

**Pull request closed without merging**